services/eventservice/src/data_manager.rs

```diff
@@ -0,0 +1,304 @@
+        let cat = new_event.category.name.as_bytes();
+        let events = self.db.open_tree(&self.events_name)?;
+        events.insert(id.clone(), new_event.to_json().unwrap().as_bytes())?;
+        let organizers = self.db.open_tree(&self.organizers_name)?;
+        let categories = self.db.open_tree(&self.categories_name)?;
+        if organizers.contains_key(&org_id)? {
+            organizers.merge(&org_id, id)?;
+        } else {
+            organizers.insert(org_id, id.to_vec())?;
+        }
+        if categories.contains_key(&cat)? {
+            categories.merge(&cat, id)?;
+        } else {
+            return Err(Box::new(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                "Unknown category",
+            )));
+        }
+        Ok(())
+    }
+
+    pub fn delete_event(&self, event_id: &u64) -> Result<(), Box<dyn std::error::Error>> {
+        let id = event_id.to_be_bytes();
+        let events = self.db.open_tree(&self.events_name)?;
+        if !events.contains_key(&id)? {
+            return Ok(());
+        }
+        let event = Event::from_json(&String::from_utf8(events.get(id)?.unwrap().to_vec())?)?;
+        let org_id = event.organizer.to_be_bytes().clone();
+        let cat = event.category.name.as_bytes();
+        events.remove(id)?;
+        let organizers = self.db.open_tree(&self.organizers_name)?;
+        let categories = self.db.open_tree(&self.categories_name)?;
+        if organizers.contains_key(&org_id)? {
+            let ids = organizers.get(&org_id)?.unwrap();
+            let new_ids = EventManager::_remove_id(&ids[..], &id);
+            if new_ids.is_some() {
+                organizers.compare_and_swap(org_id, Some(ids), new_ids)??;
+            }
+        }
+        if categories.contains_key(&cat)? {
+            let ids = categories.get(&cat)?.unwrap();
+            let new_ids = EventManager::_remove_id(&ids[..], &id);
+            if new_ids.is_some() {
+                categories.compare_and_swap(cat, Some(ids), new_ids)??;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn update_event(&self, new_event: &Event) -> Result<(), Box<dyn std::error::Error>> {
+        let id = new_event.id.to_be_bytes().clone();
+        let events = self.db.open_tree(&self.events_name)?;
+        events.insert(id, new_event.to_json().unwrap().as_bytes())?;
+        Ok(())
+    }
+
+    pub fn get_event(&self, event_id: &u64) -> Result<Event, Box<dyn std::error::Error>> {
+        let id = event_id.to_be_bytes();
+        let events = self.db.open_tree(&self.events_name)?;
+        let event = events.get(id)?;
+        if event.is_none() {
+            return Err(Box::new(std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                "Event not found",
+            )));
+        }
+        Event::from_json(&String::from_utf8(event.unwrap().to_vec())?)
+    }
+
+    pub fn get_events(
```

Comment:
please split this function. It's too huge.

Reply:
how do you know whether CRUD operations were successed? I mean please add logs like: "Event is added/updated/deleted/created". How do you handle errors received from db or when connection lost?

Reply:
I will log on upper layer, when it will be started, With question marks I'm just passing (if there are any) forward, so will display them later
When Ok(_sth_) is returned, operation was successful. If no, I will know, that sth failed, but there are no rollback mechanism yet

Reply:
And it's been stored locally, so connection won't be lost

Reply:
And according to Code Complete, optimal function size is 100-200 lines. This function is less than 70 lines
Agree, it became a bit clumsy, but I don't know how to rewrite it, without making it slower and less reliable
Also let's keep in mind, that this method is very generic one, that's also why it is bigger than the rest

Reply:
I believe we can refactor it after all.

Reply:
Just take a look what a solution I've found!

Reply:
What solution?

Reply:
Made functions for getting events for organizer and category and removed duplicated checks